File: 0230. Kth Smallest Element in a BST/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> Inorder(TreeNode* t){
        if(t == nullptr) 
            return vector<int>();
        vector<int> ret = vector<int>();
        ret.push_back(t->val);
        
        vector<int> left = vector<int>();
        vector<int> right = vector<int>();
        if(t->left)
            left = Inorder(t->left);
        if(t->right)
            right = Inorder(t->right);
        // left | ret | right
        left.insert(left.end(), ret.begin(), ret.end());
        left.insert(left.end(), right.begin(), right.end());

        return left;
        
    }
    int kthSmallest(TreeNode* root, int k) {
        vector<int> data = Inorder(root);
        return data[k-1];
    }
};

File: offer/JZ34 二叉树中和为某一值的路径(二)/tree.java
import java.util.ArrayList;
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    private void dfs(TreeNode root,int target, ArrayList<ArrayList<Integer>> ret, ArrayList<Integer> path){
        if(root  == null)
            return;
        if(root.left == null && root.right == null && root.val == target){
            ArrayList<Integer> data = new ArrayList<>(path);
            data.add(root.val);
            ret.add(new ArrayList<>(data));
            return;
        }
        path.add(root.val);
        dfs(root.left, target - root.val, ret, path);
        dfs(root.right, target - root.val, ret, path);
        path.remove(path.size() - 1);
    }
    
    public ArrayList<ArrayList<Integer>> FindPath(TreeNode root,int target) {
        ArrayList<ArrayList<Integer>> ret = new ArrayList<>();
        ArrayList<Integer> path = new ArrayList<>();
        dfs(root, target, ret, path);
        return ret;
    }
}